Add Rubocop --autocorrect functionality

maschwenk opened the pull request:
In addition to adding the `--autcorrect` feature to Rubocop Linter, this removes the dependency on Coffescript and replaces it with ES6. 

Now both the tests and src are unified as ES6 (one 💍  to rule them all), so a bit less context switching.

A bunch of build process changes, let me know @Arcanemagus if you'd prefer anything different. I can also clean up commits a bit if needed. I'm not sure if APM allows for build steps before publish? It'd be nice to not have to commit ES5 but I see other packages have the same problem (https://github.com/AtomLinter/linter-eslint/tree/master/lib)

Arcanemagus commented:
Wow, sorry, it looks like I missed this before and @AtomLinter/linter-rubocop isn't paying attention 😛.

Looking over this now.

Arcanemagus commented:
You can use ES6 (And a lot of ES2017) directly, the only reason `linter-eslint` includes transpiled code in the repo is because it launches a separate worker process which doesn't benefit from the "auto-transpilation" that Atom performs on packages.

maschwenk commented:
Thanks! One thing I actually realized I could enhance here is, for it to match how ESLint works where autocorrect tells you if it found rules it could not auto-correct. Right now occasionally the "# of fixes" is not accurate.

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+      type: WARNINGS.has(severity) ? 'Warning' : 'Error',
+      html: formatMessage({ cop_name, message, url }),
+      filePath,
+      range: [[line - 1, column - 1], [line - 1, (column + length) - 1]],
```
Do messages _always_ have a `length`?

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+const forwardRubocopToLinter =
+  ({ message: rawMessage, location, severity, cop_name }, filePath) => {
+    const { message, url } = extractUrl(rawMessage);
+    const { line, column, length } = location || DEFAULT_LOCATION;
```
If you do `helpers.generateRange(textEditor, 0)` you will get a range highlighting the entire first line. As it stands you would actually be better not specifying a range at all since the current `DEFAULT_LOCATION` will result in a range of 0 characters, which leads to it being invisible in the editor... but still there.

maschwenk replied:
I'm a little confused by this comment, could you maybe clarify a bit?

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+const formatMessage = ({ message, cop_name: copName, url }) => {
+  const formattedMessage = escapeHtml(message || DEFAULT_MESSAGE);
+  let formattedCopName = '';
+  if (copName != null && url != null) {
```
`!==`

maschwenk commented:
@Arcanemagus thanks so much for lengthy review. Will follow up with fixes!

Arcanemagus commented:
Thanks for taking the time to get this started! I was doing a check over things before starting on the ES2017 conversion myself when I saw this 😛.

Arcanemagus commented on package.json:
```diff
@@ -1,28 +1,42 @@
-  "main": "./lib/index.coffee",
-  "version": "0.5.3",
+  "main": "./src/index.js",
+  "version": "0.6.0",
```
Take the version change out, I'll do that when creating a release.

Arcanemagus replied:
Also, this will be a major version bump anyway 😛.

Arcanemagus commented on package.json:
```diff
@@ -1,28 +1,42 @@
+    "command": {
+      "type": "string",
+      "default": "rubocop",
+      "description": "\n This is the absolute path to your `rubocop` command. You may need to run \n `which rubocop` or `rbenv which rubocop` to find this. Examples: \n `/usr/local/bin/rubocop` or `/usr/local/bin/bundle exec rubocop --config \n /my/rubocop.yml`. \n"
```
Are these newlines really necessary? Atom should be handling the display of this for us.

maschwenk replied:
was going to ask you about that

Arcanemagus commented on package.json:
```diff
@@ -1,28 +1,42 @@
   "dependencies": {
-    "atom-linter": "^9.0.0",
-    "escape-html": "^1.0.3"
+    "atom-linter": "~8.0.0",
```
Let's not downgrade this 😛.

Arcanemagus commented on spec/linter-rubocop-spec.js:
```diff
@@ -1,14 +1,17 @@
-const goodPath = path.join(__dirname, 'fixtures', 'good.rb');
-const invalidWithUrlPath = path.join(__dirname, 'fixtures', 'invalid_with_url.rb');
-const invalidWithoutUrlPath = path.join(__dirname, 'fixtures', 'invalid_without_url.rb');
+const dirname = __dirname;
```
Any reason for pulling this into a new variable?

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,127 @@
+            return atom.notifications.addError('Linter-Rubocop: Please save before fixing');
+          }
+
+          const cwd = path.dirname(helpers.find(filePath, '.'));
```
I still don't see the reasoning behind this use of `helpers.find`. Why isn't this just `path.dirname(filePath)`?

maschwenk commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+    const config = helpers.find(filePath, '.rubocop.yml');
+    if (config === null) { return []; }
+  }
+  const cwd = path.dirname(helpers.find(filePath, '.'));
```
This was the implementation in the original coffescript. What would you suggest instead? `__dirname`?

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,127 @@
+const forwardRubocopToLinter =
+  ({ message: rawMessage, location, severity, cop_name }, filePath) => {
+    const [message, url] = rawMessage.split(/ \((.*)\)/, 2);
+    const { line, column, length } = location || { line: 1, column: 1, length: 0 };
```
Again, if `location` isn't specified you would be better not specifying a `range` attribute at all on the message since this default will result in a zero length range which is invisible in the editor (but still "there"!).

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,127 @@
+          const config = helpers.find(filePath, '.rubocop.yml');
+          if (config === null) { return []; }
+        }
+        const cwd = path.dirname(helpers.find(filePath, '.'));
```
Again, this `helpers.find()` usage is just plain bizarre and I don't see how it's any different from `path.dirname(filePath)`.

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,127 @@
+                            .split(/\s+/)
+                            .filter(i => i)
+                            .concat(DEFAULT_ARGS, '--stdin', filePath);
+        if (this.disableWhenNoConfigFile === true) {
```
Might as well move this above the `command` parsing, since if this check fails there is no need to do that work (little as it may be).

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,127 @@
+                            .concat(DEFAULT_ARGS, '--stdin', filePath);
+        if (this.disableWhenNoConfigFile === true) {
+          const config = helpers.find(filePath, '.rubocop.yml');
+          if (config === null) { return []; }
```
Can you put the code block on new lines?

Arcanemagus commented on package.json:
```diff
@@ -1,28 +1,42 @@
-    "escape-html": "^1.0.3"
+    "escape-html": "^1.0.3",
+    "pluralize": "^3.1.0",
+    "tmp": "0.0.31"
```
Any reason to pin this? Greenkeeper will trigger a build for all new versions checking if it causes any issues.

Arcanemagus commented:
Btw, would you like to be added as a maintainer here? It doesn't look like the current member of @AtomLinter/linter-rubocop is responding, and you are doing a complete re-write here.

caseywebdev commented:
I'm not really invested in this particular project any more (not currently working with ruby) so by all means take the reins.

maschwenk commented:
yeah that'd be awesome! Just addressed all your comments I believe, itemized them with `git add -p` but will rebase/clean all these commits once you give :shipit:

maschwenk commented:
btw @Arcanemagus it's been really helpful looking through all the Linters you maintain, lots of cool tricks to learn

maschwenk commented on src/index.js:
```diff
@@ -39,13 +39,15 @@ const parseFromStd = (stdout, stderr) => {
       filePath,
-      range: [[line - 1, column - 1], [line - 1, (column + length) - 1]],
     };
+    return Object.assign(outputToLinter,
```
I'm iffy about this, let me know if it's 👎

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,134 @@
+      filePath,
+    };
+    return Object.assign(outputToLinter,
+      location && { range: [[line - 1, column - 1], [line - 1, (column + length) - 1]] },
```
Although this is nice and compact, wouldn't this throw a `TypeError` if location was undefined (since it would then be `false` and not an `Object`)?

I was thinking something simpler, along the lines of:

```js
if (location) {
  outputToLinter.range = [[line - 1, column - 1], [line - 1, (column + length) - 1]];
}
return outputToLinter;
```

maschwenk replied:
It actually just returns itself apparently:

![image](https://cloud.githubusercontent.com/assets/1264305/23522777/2675f368-ff39-11e6-8154-a3ecdfe67b3a.png)

I'm cool with either way really. It's such a common pattern I see in my other projects, I was interested to find that this works.

Arcanemagus replied:
Huh, cool. I'm fine with it then.

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,134 @@
+                              .filter(i => i)
+                              .concat(DEFAULT_ARGS, '--auto-correct', filePath);
+
+          if (!atom.workspace.isTextEditor(textEditor) || textEditor.isModified()) {
```
Should probably move this up to right after getting `textEditor`.

Arcanemagus commented on src/index.js:
```diff
@@ -0,0 +1,134 @@
+        const filePath = editor.getPath();
+
+        if (this.disableWhenNoConfigFile === true) {
+          const config = helpers.find(filePath, '.rubocop.yml');
```
This can be changed to `findAsync` and `await`'d on.

Arcanemagus commented:
Added you to the team @maschwenk 🎉.

Arcanemagus commented:
Looks good to me at this point, feel free to merge!

As far as releasing this you can either use `apm publish {patch | minor | major}` to have APM do everything for you, or if you want a bit more control (like having the release be signed) I have a script that does the following that I use myself:

```bash
# ... setting $version, and verifying that there are staged changes
git commit --gpg-sign --message="Prepare $version release"
git tag --sign --message="$version" "$version"
git push --follow-tags
apm publish --tag "$version"
```

Before running that I stage the changelog and `package.json` changes.
